### PR TITLE
planner: update some cost formulas of model ver2

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #35240

Problem Summary: planner: update some cost formulas of model ver2

### What is changed and how it works?

1. update the cost formula of Sort/TopN from `rows*log(rows)*cols*cpu_factor` to `rows*cols*cpu_factor + rows*log(rows)*cpu_factor`;
2. consider `BCastJoin` and `ShuffleJoin` in `Exchange` and `HashJoin`; 

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by xuyifangreeneyes — APPROVED** [review]

File: planner/core/plan_cost_ver2.go [hunk 1]
```diff
@@ -510,8 +509,13 @@ func (p *PhysicalHashJoin) getPlanCostVer2(taskType property.TaskType, option *P
 		return zeroCostVer2, err
 	}
 
-	p.planCostVer2 = sumCostVer2(buildChildCost, probeChildCost, buildHashCost, buildFilterCost,
-		divCostVer2(sumCostVer2(probeFilterCost, probeHashCost), concurrency))
+	if taskType == property.MppTaskType { // BCast or Shuffle Join, use mppConcurrency
+		p.planCostVer2 = sumCostVer2(buildChildCost, probeChildCost,
+			divCostVer2(sumCostVer2(buildHashCost, buildFilterCost, probeHashCost, probeHashCost), mppConcurrency))
```
Comment (time-and-fate) [inline planner/core/plan_cost_ver2.go:514]: Is this expected? The formula is changed.

**Review by time-and-fate — COMMENTED** [review]
Why do we need to change counting `Expression`s into counting `ScarlarFunction`s?

**Comment by qw4990** [comment]
> Why do we need to change counting `Expression`s into counting `ScarlarFunction`s?

Since currently the formula `rows * cols * cpu-factor` is used to represent how much calculation TiDB has to do, if it's not a `ScarlarFunction` we'd better ignore it here.

File: planner/core/plan_cost_ver2.go (hunk 1 above)
Comment (time-and-fate) [reply planner/core/plan_cost_ver2.go:514]: Ok.
But the `probeHashCost` appeared two times.

Comment (qw4990) [reply planner/core/plan_cost_ver2.go]: My fault. It's updated, PTAL @time-and-fate

**Comment by time-and-fate** [comment]
> > Why do we need to change counting `Expression`s into counting `ScarlarFunction`s?
> 
> Since currently the formula `rows * cols * cpu-factor` is used to represent how much calculation TiDB has to do, if it's not a `ScarlarFunction` we'd better ignore it here.

I think if it's not a `ScalarFunction` here, it also needs "calculation".
Like `order by a` or `t1 left join t2 on t1.a`, I think they are not much different from `order by a+1` or `t1 left join t2 on t1.a>10`.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**